nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,372 @@
+#include <nano/node/confirmation_height_processor.hpp>
+
+#include <boost/optional.hpp>
+#include <cassert>
+#include <nano/lib/logger_mt.hpp>
+#include <nano/lib/numbers.hpp>
+#include <nano/lib/utility.hpp>
+#include <nano/node/active_transactions.hpp>
+#include <nano/node/stats.hpp>
+#include <nano/secure/blockstore.hpp>
+#include <nano/secure/common.hpp>
+#include <nano/secure/ledger.hpp>
+#include <numeric>
+
+nano::confirmation_height_processor::confirmation_height_processor (nano::pending_confirmation_height & pending_confirmation_height, nano::block_store & store, nano::ledger & ledger, nano::active_transactions & active, nano::logger_mt & logger) :
+pending_confirmations (pending_confirmation_height),
+store (store),
+ledger (ledger),
+active (active),
+logger (logger),
+thread ([this]() {
+	nano::thread_role::set (nano::thread_role::name::confirmation_height_processing);
+	this->run ();
+})
+{
+}
+
+nano::confirmation_height_processor::~confirmation_height_processor ()
+{
+	stop ();
+}
+
+void nano::confirmation_height_processor::stop ()
+{
+	stopped = true;
+	condition.notify_one ();
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+void nano::confirmation_height_processor::run ()
+{
+	std::unique_lock<std::mutex> lk (pending_confirmations.mutex);
+	while (!stopped)
+	{
+		if (!pending_confirmations.pending.empty ())
+		{
+			current_original_pending_block = *pending_confirmations.pending.begin ();
+			pending_confirmations.pending.erase (current_original_pending_block);
+			lk.unlock ();
+			add_confirmation_height (current_original_pending_block);
+			current_original_pending_block = 0;
+			lk.lock ();
+		}
+		else
+		{
+			condition.wait (lk);
+		}
+	}
+}
+
+void nano::confirmation_height_processor::add (nano::block_hash const & hash_a)
+{
+	{
+		std::lock_guard<std::mutex> lk (pending_confirmations.mutex);
+		pending_confirmations.pending.insert (hash_a);
+	}
+	condition.notify_one ();
+}
+
+// This only check top-level blocks having their confirmation height sets, not anything below
+bool nano::confirmation_height_processor::is_processing_block (nano::block_hash const & hash_a)
+{
+	// First check the hash currently being processed
+	if (!current_original_pending_block.is_zero () && current_original_pending_block == hash_a)
```

Comment:
Same question about locking here (though I think `is_processing_block` is only used in tests)

Reply:
This is also used in the node. I have protected it now (moved the logic to the `pending_confirmation_height` class)